### PR TITLE
Implement PE with deferred intent for the card payment method in the classic checkout

### LINKED ISSUE #2747: [Deferred intent] Card - Process payments in the classic checkout
### Acceptance criteria
- No intent must be created when visiting the checkout page
- The WC Order must contain the same metadata as when using the legacy PE:
  -  _stripe_charge_captured
  - _stripe_currency
  - _stripe_fee
  - _stripe_intent_id
  - _stripe_net
  - _stripe_upe_payment_type
  - The meta _stripe_upe_redirect_processed must not be set as it's unnecessary. No processing on redirect takes place with the deferred intent

- The Payment intent in Stripe must contain the same information as it does with the legacy PE. Check:
  - Customer data (both when checking out as a guest and a logged-in user)
  - Payment details
    - For the statement descriptor, it must reflect the short or long version depending on the options in the plugin dashboard
  - Payment method data 
  - Metadata:
    - customer_email
    - customer_name
    - order_id
    - order_key
    - payment_type
    - site_url

- Errors must be handled:
  - A relevant, user-friendly message must be shown to shoppers
  - A relevant note must be added to the order
  - A detailed message of the error must be logged to the log files if enabled
  - The order status must reflect the payment status

- Must be covered by tests
- The Legacy PE code that will be replaced by this implementation must be removed if it's not used somewhere else
- The following critical flows and e2e tests must behave as expected (use the legacy PE implementation as reference if in doubt):
  - [Shopper > Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#checkout-with-normal-credit-card)
  - [Shopper > Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#checkout-failures-with-various-cards)

### Out of scope
These will be handled in separate issues to keep the initial PR as small as possible:

- Implementation on the Block checkout
- Using a saved payment method
- Reusing an already created payment intent
- Using setup intents
- Handling authorizing transactions and capturing later
- Alternative payment methods
- Compatibility with Pre-Orders and Subscriptions

### PR DESCRIPTION
**Pull request by a-danae**
Fixes #2747

This PR doesn't point to `develop` as we'll require further work before these changes get shipped.

## Changes proposed in this Pull Request:

- Collect the payment details before creating the payment intent (aka, use [deferred intents](https://stripe.com/docs/payments/accept-a-payment-deferred)). Previously, a payment intent was created when the checkout page was loaded.
- When processing the payment, the payment intent gets created already with the payment details and confirmed.
- Processing UPE payments after the redirection to the "Order confirmation" page isn't needed anymore.
- When the payment processing fails, the failed response gets added to the debug log

This PR only covers:
  - Processing payments from the classic checkout page. The block checkout, Pay for order page, and adding a new payment method from my account aren't covered.
  - Automatic captures. Authorizing the transaction and capturing later will be handled separately.
  - Using new cards regular. Other payment methods, 3DS cards, saving a card, or reusing an already saved card will be handled separately.

## Testing instructions

- Enable UPE
- As a shopper, process payments under the classic checkout page for all tests

**Happy path**
- Check out `develop` 
- Pay for an order using a new regular card, not saving it. Flow: [Shopper > Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#checkout-with-normal-credit-card)
- Check out this branch and pay for another order
- Open both WC orders in the database
- Confirm these meta exist and match between both orders:
  -  _stripe_charge_captured
  - _stripe_currency
  - _stripe_fee
  - _stripe_intent_id
  - _stripe_net
  - _stripe_upe_payment_type
- Confirm that the meta _stripe_upe_redirect_processed, isn't set in the order processed using this branch
- Open the payment intents in Stripe for both orders
- Confirm that the information for the payment intents created for each order is consistent with each other:
  - Customer data. You could try checking out as:
    - A guest -> A new Stripe customer should be created
    - A customer with no associated Stripe user -> A new Stripe customer should be created. It'll be reused
    - A customer with an associated Stripe user -> No new Stripe customer should be created. The associated one should be reused
  - Payment details
    - For the statement descriptor, it must reflect the short or long version depending on the options in the plugin dashboard
  - Payment method data 
  - Metadata:
    - customer_email
    - customer_name
    - order_id
    - order_key
    - payment_type
    - site_url

**Confirm the intent creation flow is deferred**
- Add a product to the cart and go to the checkout page. Don't continue
- In the Stripe dashboard, confirm no payment intent was created
- Proceed with the checkout
- Confirm that the payment intent was created and confirmed

**Card failures**
  - Confirm the following flow behaves as expected [Shopper > Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#checkout-failures-with-various-cards)

**Note:** With the card failures, a new payment intent will be created for each retry. In the future, the same intent will be reused as possible. This will be fixed by https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2761.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 1]
```diff
@@ -503,6 +519,11 @@ public function payment_fields() {
 	 * @return array|null An array with result of payment and redirect URL, or nothing.
 	 */
 	public function process_payment( $order_id, $retry = true, $force_save_source = false, $previous_error = false, $use_order_source = false ) {
+		// Flag for using a deferred intent. To be removed.
+		if ( ! empty( $_POST['wc-stripe-is-deferred-intent'] ) ) {
```
Comment (a-danae) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:522]: I added this to avoid interfering with other flows, like the block checkout one. And also to keep the old logic around until we're not using it anywhere, and to keep the new logic separated to make the cleanup easier later. Glad to handle this differently, tho.

File: client/classic/upe/payment-processing.js [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+	}
+
+	return api
+		.getStripe( paymentMethodType )
+		.createPaymentMethod( { elements, params } );
+}
+
+/**
+ * Mounts the existing Stripe Payment Element to the DOM element.
+ * Creates the Stipe Payment Element instance if it doesn't exist and mounts it to the DOM element.
```
Comment (mattallan) [inline client/classic/upe/payment-processing.js]:
```suggestion
 * Creates the Stripe Payment Element instance if it doesn't exist and mounts it to the DOM element.
```

Small typo

File: client/classic/upe/payment-processing.js [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+	 *
+	 * Ref: https://github.com/woocommerce/woocommerce/blob/2429498/assets/js/frontend/tokenization-form.js#L109
+	 */
+	const event = new Event( 'wc-credit-card-form-init' );
+	document.body.dispatchEvent( event );
+
+	const paymentMethodType = domElement.dataset.paymentMethodType;
+	let upeElement;
+
+	// Non-split PE. To be removed.
```
Comment (mattallan) [inline client/classic/upe/payment-processing.js:212]: > To be removed.

Just making sure I understand this and other similar in-line comments. This section can be removed when we stop supporting the legacy PE?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 4]
```diff
@@ -1489,4 +1577,86 @@ private function get_address_data_for_payment_request( $order ) {
...
 	}
 
+	/**
+	 * Collects the payment information needed for processing a payment intent.
+	 *
+	 * @param WC_Order $order The WC Order to be paid for.
+	 * @return array An array containing the payment information for processing a payment intent.
+	 */
+	private function prepare_payment_information_from_request( WC_Order $order ) {
+		// TODO: throw exception if any required information is missing.
```
Comment (mattallan) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Just checking, are the TODO's added in this PR going to be addressed in this PR or later?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 5]
```diff
@@ -1489,4 +1577,86 @@ private function get_address_data_for_payment_request( $order ) {
...
+
+		$payment_method               = sanitize_text_field( wp_unslash( $_POST['wc-stripe-payment-method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+		$selected_payment_type        = sanitize_text_field( wp_unslash( $_POST['wc_stripe_selected_upe_payment_type'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+		$capture_method               = empty( $this->get_option( 'capture' ) ) || $this->get_option( 'capture' ) === 'yes' ? 'automatic' : 'manual'; // automatic | manual.
+		$save_payment_method_to_store = isset( $_POST['save_payment_method'] ) ? 'yes' === wc_clean( wp_unslash( $_POST['save_payment_method'] ) ) : false;
+		$currency                     = strtolower( get_woocommerce_currency() );
```
Comment (mattallan) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]:
```suggestion
		$currency                     = strtolower( $order->get_currency() );
```

I think this should ise the orders currency not the stores currency, is that correct?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 1 above)
Comment (mattallan) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:522]: I like it! :)

**Review by mattallan — COMMENTED** [review]
Niiiice changes @a-danae !! I'm excited to see the work being started in this PR.

I've only just performed a quick code review so far but I'll have play with the front-end changes and then look over the changes with more attention to detail. 🙌

File: includes/class-wc-stripe-intent-controller.php [hunk 6]
```diff
@@ -676,6 +676,152 @@ public function maybe_process_upe_redirect() {
...
+			$payment_information['level3'],
+			$order
+		);
+
+		// Only update the payment_type if we have a reference to the payment type the customer selected.
+		if ( '' !== $selected_payment_type ) {
+			$order->update_meta_data( '_stripe_upe_payment_type', $selected_payment_type );
+		}
+
+		$order->update_status( 'pending', __( 'Awaiting payment.', 'woocommerce-gateway-stripe' ) );
```
Comment (mattallan) [inline includes/class-wc-stripe-intent-controller.php]: I haven't tested anything yet but I'm curious why we need to update the status of the order to pending when dealing with deferred intents. What is the order status before this?

I'm wondering if we should have a `$order->needs_payment()` somewhere so that we're only attempting to create and confirm an intent for orders that have pending/failed status.

Maybe inside `process_payment_with_deferred_intent()`?

File: client/classic/upe/payment-processing.js (hunk 2 above)
Comment (a-danae) [reply client/classic/upe/payment-processing.js]: Whooops. Thanks! Committed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/19030899609fcca977ec55cdc56b3da1aea1f03b

File: client/classic/upe/payment-processing.js (hunk 3 above)
Comment (a-danae) [reply client/classic/upe/payment-processing.js:212]: > This section can be removed when we stop supporting the legacy PE?

Exactly. More specifically, this can be removed when we implement using split PE. 

We could fully switch to using deferred intents, but we'll still need this until we split the payment elements in the frontend in separate WooCommerce gateways.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 4 above)
Comment (a-danae) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: This one specifically was already solved. Removed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/8098d581ea806a23d76e5afe43b160793f2a9f58

But, yup, it's probably confusing to have them around and we already have tasks for working on them. I'm removing the ones that aren't meant to be solved in this PR, which is all but one.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 5 above)
Comment (a-danae) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Nice catch! Yes, I think you're right.

I followed the [old implementation](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-intent-controller.php#L338) when creating an intent where the store currency is used, but we didn't have an order at that point with legacy PE. But we do have an order in this case.

Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/3c4fb1e7433a87559bf51b48ea4db705761d4810. Thanksss

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 4 above)
Comment (mattallan) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Thanks @a-danae !

> It's probably confusing to have them around and we already have tasks for working on them

I personally think they're fine to have in this base PR to help indicate where we expect code from another issue to go. I just wanted to avoid shipping the feature with a bunch of TODO's that we added early on and then never got around to addressing 😄 

Happy to go with whatever you think is best!

File: includes/class-wc-stripe-intent-controller.php (hunk 6 above)
Comment (a-danae) [reply includes/class-wc-stripe-intent-controller.php]: > I haven't tested anything yet but I'm curious why we need to update the status of the order to pending when dealing with deferred intents. What is the order status before this?

You're right, no need to update the order status. The order is already "pending" at this point. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/fef7018fddf43487b8205e46ed3ad4145e8ae289

> I'm wondering if we should have a $order->needs_payment() somewhere so that we're only attempting to create and confirm an intent for orders that have pending/failed status.

Yeah, I agree. 

I see there's a more complex logic around this in [verify_intent_after_checkout()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-gateway-stripe.php#L823), but I think it's specific for the legacy PE implementation because it runs between the Checkout and Thank you pages.

Would it be okay with you to have an issue for addressing this, instead of doing it in this PR? I'm thinking that webhooks and cache may also come into play, but would like to dive into it. Also to not delay merging this base PR.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 4 above)
Comment (a-danae) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Makes complete sense 😄 

I've removed the TODO comments in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/cf90c3c838baae55eab87de5098239b27f443fd1. We can decide where's the best place to handle the behaviors mentioned in those comments when working on the issues.

File: includes/class-wc-stripe-intent-controller.php (hunk 6 above)
Comment (mattallan) [reply includes/class-wc-stripe-intent-controller.php]: > Would it be okay with you to have an issue for addressing this, instead of doing it in this PR? I'm thinking that webhooks and cache may also come into play, but would like to dive into it. Also to not delay merging this base PR.

That sounds good to me!

File: client/classic/upe/payment-processing.js [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+/**
+ * Initializes the appearance of the payment element by retrieving the UPE configuration
+ * from the API and saving the appearance if it doesn't exist. If the appearance already exists,
+ * it is simply returned.
+ *
+ * @return {Object} The appearance object for the UPE.
+ */
+function initializeAppearance() {
```
Comment (mattallan) [inline client/classic/upe/payment-processing.js:32]: Just noticed this was empty, will this be implemented in a separate PR?

**Review by mattallan — APPROVED** [review]
Amazing work on this one @a-danae !!

As per the testing instructions you shared on the PR, I purchased two orders using a new card on both the `develop` and `add/card-classic-with-deferred-intent` and ran through the following checks I've posted below.

The only two issues I found were when I tried to purchase a regular WC product as a guest customer and the other issue was I noticed shipping details were missing from the payment data when looking at the payment from the Stripe dashboard

I'm not sure if you wanted to fix these in this PR or open separate issues for them but I'm happy to approve this PR and leave that decision up to you 😄 

I also left a small comment regarding the empty `initializeAppearance` function

---

- [x] Confirmed the `_stripe_upe_redirect_processed` meta is not set on the order purchased on this branch
- [x] Confirmed all of the stripe processing meta is set on the order (i.e. `_stripe_fee` & `_stripe_intent_id`) and is consistent across both orders
- [x] Confirmed the payment intent in the Stripe dashabord has the correct customer, payment & meta data and is consistent across both orders created on different branched.
    - ⚠️ The customer, payment and metadata all looked correct, however one thing I noticed is the payment on this branch doesn't have a shipping section when looking at Stripe. Looking in the events and logs section, it looks like on `develop` we send a `POST /payment_intents/pi_` request with shipping. Not sure if this is needed but I'm happy to move this to a separate issue if you like.
image


Tested the above cases as:
- [ ] Guest customer
   - 🚨 I ran into an error trying to purchase a product as a guest customer on this branch. There weren't any PHP errors or gateway/request logs but I noticed this error in the console.
image
   - [x] Existing WC customer without Stripe customer
   - [x] Existing WC customer with Stripe csutomer
- [x] Tested with a failed card (0431) along with other erroring cases and confirmed new intents were created each time on this branch (confirming the bug that will be fixed in #2761)

Specific deferred intent checks:
- [x] Confirm no intent is created on checkout page load on this branch
- [x] Confirm intent is created and confirmed after processing the checkout

File: includes/class-wc-stripe-intent-controller.php (hunk 6 above)
Comment (a-danae) [reply includes/class-wc-stripe-intent-controller.php]: Yay! GH issue created over here https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2785 . Of course feel free to update it!

And thanks as always for pointing these out 😄

File: client/classic/upe/payment-processing.js (hunk 7 above)
Comment (a-danae) [reply client/classic/upe/payment-processing.js:32]: Thanks! I meant to do it in this PR 🙂 fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/66977492f0ed1ee986675d50f87b2297132610b3

Comment (mattallan) [reply client/classic/upe/payment-processing.js:32]: Thanks!

**Comment by mattallan** [comment]
Thanks for the latest changes @a-danae !

I've confirmed [61d9e34](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/61d9e34c711a13ab492cb61da6f76d632ec5728a) fixes the issue with shipping details not being sent to Stripe. Here's a screenshot testing the latest:

image

💃

**Comment by a-danae** [comment]
Thanks for going ahead and re-reviewing this PR! I completely missed the missing shipping information, great catch 🙂  I was planning to leave a comment here after solving the guest customer issue but I'll continue with that in the morning.

**Comment by a-danae** [comment]
> Guest customer
🚨 I ran into an error trying to purchase a product as a guest customer on this branch. There weren't any PHP errors or gateway/request logs but I noticed this error in the console.

This should be fixed by https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/863cfc790eb88b12d2b496c997e894ea1334002e. Something similar was happening in another place, and should be fixed by https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/1c2f3089869c99e36d1240f3e4f014efcd935759

Also, I introduced this validation that was present in the legacy PE flow https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/646299c4eefd68f302f7c82830729f74beca654a

Back to you. Thanks for such a thorough review, Matt! 🙌

**Review by mattallan — APPROVED** [review]
Thanks @a-danae for fixing those issues!

- [x] I've confirmed the issue with not being able to purchase as a guest customer have been resolved.
- [x] I've given the latest changes a code review
- [x] Smoke tested the overall changes once again

I'm happy to approve this to be merged 🎉💃 
Thanks for all the work you did to lay the foundation for deferred intents!
